File: src/normalize/rank.ts
import type { RawSearchResult, SearchResult } from "../providers/types.js";
import { extractDomain } from "./dedup.js";

// Domain authority tiers (higher = more authoritative)
const DOMAIN_AUTHORITY: Record<string, number> = {
  "github.com": 0.95,
  "stackoverflow.com": 0.95,
  "developer.mozilla.org": 0.95,
  "docs.python.org": 0.90,
  "docs.microsoft.com": 0.90,
  "learn.microsoft.com": 0.90,
  "wikipedia.org": 0.85,
  "en.wikipedia.org": 0.85,
  "arxiv.org": 0.85,
  "medium.com": 0.60,
  "dev.to": 0.65,
  "hackernews.com": 0.70,
  "news.ycombinator.com": 0.75,
  "reddit.com": 0.60,
  "twitter.com": 0.50,
  "x.com": 0.50,
};

const DOMAIN_PATTERN_AUTHORITY: [RegExp, number][] = [
  [/\.gov$/, 0.90],
  [/\.edu$/, 0.85],
  [/\.org$/, 0.70],
  [/docs\./, 0.80],
  [/wiki\./, 0.70],
];

function getDomainAuthority(domain: string): number {
  if (DOMAIN_AUTHORITY[domain] !== undefined) {
    return DOMAIN_AUTHORITY[domain];
  }
  for (const [pattern, score] of DOMAIN_PATTERN_AUTHORITY) {
    if (pattern.test(domain)) {
      return score;
    }
  }
  return 0.50; // default
}

function computeKeywordMatch(query: string, title: string, snippet: string): number {
  const queryTerms = query.toLowerCase().split(/\s+/).filter((t) => t.length > 2);
  if (queryTerms.length === 0) return 0.5;

  const text = `${title} ${snippet}`.toLowerCase();
  let matches = 0;
  for (const term of queryTerms) {
    if (text.includes(term)) matches++;
  }
  return matches / queryTerms.length;
}

function computeFreshness(publishedDate: string | null): number {
  if (!publishedDate) return 0.5; // neutral when unknown
  try {
    const date = new Date(publishedDate);
    const now = new Date();
    const daysSince = (now.getTime() - date.getTime()) / (1000 * 60 * 60 * 24);
    if (daysSince < 7) return 1.0;
    if (daysSince < 30) return 0.85;
    if (daysSince < 90) return 0.70;
    if (daysSince < 365) return 0.55;
    return 0.30;
  } catch {
    return 0.5;
  }
}

interface RankingContext {
  query: string;
  engineCounts: Map<string, Set<string>>; // url -> set of engines that returned it
}

const WEIGHTS = {
  position: 0.40,
  keyword: 0.25,
  authority: 0.20,
  freshness: 0.15,
};

const CROSS_ENGINE_BOOST = 1.2;

export function rankResults(
  raw: RawSearchResult[],
  query: string,
): SearchResult[] {
  // Track which engines returned each URL
  const engineSets = new Map<string, Set<string>>();
  for (const r of raw) {
    const key = r.url.toLowerCase();
    if (!engineSets.has(key)) engineSets.set(key, new Set());
    engineSets.get(key)!.add(r.engine);
  }

  // Score each result
  const scored: SearchResult[] = raw.map((r, index) => {
    const domain = extractDomain(r.url);
    const positionScore = 1 / (index + 1);
    const keywordScore = computeKeywordMatch(query, r.title, r.snippet);
    const authorityScore = getDomainAuthority(domain);
    const freshnessScore = computeFreshness(r.publishedDate);

    let relevanceScore =
      WEIGHTS.position * positionScore +
      WEIGHTS.keyword * keywordScore +
      WEIGHTS.authority * authorityScore +
      WEIGHTS.freshness * freshnessScore;

    // Boost results that appear in multiple engines
    const engines = engineSets.get(r.url.toLowerCase());
    if (engines && engines.size > 1) {
      relevanceScore *= CROSS_ENGINE_BOOST;
    }

    // Clamp to [0, 1]
    relevanceScore = Math.min(1, Math.max(0, relevanceScore));

    return {
      rank: 0, // assigned after sorting
      title: r.title,
      url: r.url,
      snippet: r.snippet,
      engine: r.engine,
      publishedDate: r.publishedDate,
      domain,
      relevanceScore: Math.round(relevanceScore * 1000) / 1000,
      summary: null,
    };
  });

  // Sort by relevanceScore descending
  scored.sort((a, b) => b.relevanceScore - a.relevanceScore);

  // Assign ranks
  for (let i = 0; i < scored.length; i++) {
    scored[i].rank = i + 1;
  }

  return scored;
}

File: src/evoscry/normalize/rank.py
"""Relevance ranking with 4-factor weighted scoring."""

from __future__ import annotations

import re
from datetime import datetime, timezone

from evoscry.normalize.dedup import extract_domain

# Domain authority tiers
DOMAIN_AUTHORITY: dict[str, float] = {
    "github.com": 0.95,
    "stackoverflow.com": 0.95,
    "developer.mozilla.org": 0.95,
    "docs.python.org": 0.90,
    "docs.microsoft.com": 0.90,
    "learn.microsoft.com": 0.90,
    "wikipedia.org": 0.85,
    "en.wikipedia.org": 0.85,
    "arxiv.org": 0.85,
    "medium.com": 0.60,
    "dev.to": 0.65,
    "news.ycombinator.com": 0.75,
    "reddit.com": 0.60,
    "twitter.com": 0.50,
    "x.com": 0.50,
}

DOMAIN_PATTERNS: list[tuple[re.Pattern, float]] = [
    (re.compile(r"\.gov$"), 0.90),
    (re.compile(r"\.edu$"), 0.85),
    (re.compile(r"\.org$"), 0.70),
    (re.compile(r"^docs\."), 0.80),
    (re.compile(r"^wiki\."), 0.70),
]

# Scoring weights
W_POSITION = 0.40
W_KEYWORD = 0.25
W_AUTHORITY = 0.20
W_FRESHNESS = 0.15
CROSS_ENGINE_BOOST = 1.2


def _get_domain_authority(domain: str) -> float:
    if domain in DOMAIN_AUTHORITY:
        return DOMAIN_AUTHORITY[domain]
    for pattern, score in DOMAIN_PATTERNS:
        if pattern.search(domain):
            return score
    return 0.50


def _keyword_match(query: str, title: str, snippet: str) -> float:
    terms = [t for t in query.lower().split() if len(t) > 2]
    if not terms:
        return 0.5
    text = f"{title} {snippet}".lower()
    matches = sum(1 for t in terms if t in text)
    return matches / len(terms)


def _freshness_score(published_date: str | None) -> float:
    if not published_date:
        return 0.5
    try:
        d = datetime.strptime(published_date, "%Y-%m-%d").replace(tzinfo=timezone.utc)
        days = (datetime.now(timezone.utc) - d).days
        if days < 7:
            return 1.0
        if days < 30:
            return 0.85
        if days < 90:
            return 0.70
        if days < 365:
            return 0.55
        return 0.30
    except Exception:
        return 0.5


def rank_results(raw: list[dict], query: str) -> list[dict]:
    """Score and sort results by relevance. Returns list of dicts with 'score' added."""
    # Track which engines returned each URL
    engine_sets: dict[str, set[str]] = {}
    for r in raw:
        key = r["url"].lower()
        engine_sets.setdefault(key, set()).add(r.get("engine", "unknown"))

    scored = []
    for i, r in enumerate(raw):
        domain = extract_domain(r["url"])
        position = max(0, 1.0 - (i / max(len(raw), 1)))
        keyword = _keyword_match(query, r.get("title", ""), r.get("snippet", ""))
        authority = _get_domain_authority(domain)
        freshness = _freshness_score(r.get("published_date"))

        score = (
            W_POSITION * position
            + W_KEYWORD * keyword
            + W_AUTHORITY * authority
            + W_FRESHNESS * freshness
        )

        # Cross-engine boost
        key = r["url"].lower()
        if len(engine_sets.get(key, set())) > 1:
            score *= CROSS_ENGINE_BOOST

        result = dict(r)
        result["score"] = round(score, 4)
        result["domain"] = domain
        # Drop internal tracking fields
        result.pop("_engines", None)
        scored.append(result)

    scored.sort(key=lambda r: r["score"], reverse=True)
    return scored
